### PR TITLE
CLOUDP-32118: (Lawrence, Chunming) Track HTTP download progress for each request

### PR DESCRIPTION
**Pull request by kaywux**
https://jira.mongodb.org/browse/CLOUDP-32118

Add tracking functionality to backup restore job progress: sum up total bytes to transfer, keep track of bytes currently transferred by each client.

QA Performed:
Tested on local machine by creating large files and transferring them through curl statement

### CODE REVIEW

File: atlas_restore.go [hunk 1]
```diff
@@ -35,7 +35,30 @@ type Config struct {
...
+	// temp holds number of bytes streamed since last push, total bytes streamed
+	// is only pushed once temp reaches threshold (currently 500 megabytes)
+	if writer.temp > 500000000 {
+		writer.tracker.BytesStreamed(writer.total)
+		writer.temp = 0
+	}
+	return n, err
+}
+
+func addToTar(tw *tar.Writer, pw *ProgressWriter, dbPath, tarParentDir, relPath string, excludePaths []string) error {
```
Comment (lawctan) [inline atlas_restore.go]: we don't need pw in here, passing the pw object to the 1st argument for tw will also work. (I think)

File: config.yaml [hunk 2]
```diff
@@ -1,15 +1,15 @@
-authKey:
-filename:
-tarDirectory:
-dbPath:
-downloadMode: true
-port:
-ssl: true
+authKey: SUqxnHjwQBVdOnFFT5kJ8mdE
```
Comment (lawctan) [inline config.yaml]: let's not commit these changes, we can probably add this to .gitignore

File: status.go [hunk 3]
```diff
@@ -10,13 +10,15 @@ import (
 )
 
 type DownloadStats struct {
-	InProgress int32 `json:"inProgress"`
-	Failed     int32 `json:"failed"`
-	Succeeded  int32 `json:"succeeded"`
+	InProgress		int32 `json:"inProgress"`
+	Failed     		int32 `json:"failed"`
+	Succeeded  		int32 `json:"succeeded"`
+	BytesStreamed 	int64 `json:"bytesStreamed"`
```
Comment (lawctan) [inline status.go]: not sure what's the naming convention and why the fields are capitalized. Also, this needs to be changed to consider multiple clients

File: status.go [hunk 4]
```diff
@@ -30,19 +32,17 @@ type StatsTracker struct {
 	client     *http.Client
 	jobId      string
 	authKey    string
-	shouldPush bool
```
Comment (lawctan) [inline status.go]: i think `shouldPush` is better for abstraction. Using download creates some assumption about implementation

File: status.go [hunk 5]
```diff
@@ -30,19 +32,17 @@ type StatsTracker struct {
...
 
-func NewStatsTracker(centralUrl string, jobId string, authKey string, shouldPush bool) *StatsTracker {
+func NewStatsTracker(centralUrl string, jobId string, authKey string, downloadMode bool) *StatsTracker {
 	tracker := &StatsTracker{
 		stats:      DownloadStats{},
 		centralUrl: centralUrl,
 		jobId:      jobId,
 		authKey:    authKey,
-	}
-	if shouldPush {
```
Comment (lawctan) [inline status.go]: i think we should keep the shouldPush, we don't wanna push info when doing manual download.

File: atlas_restore.go [hunk 6]
```diff
@@ -35,7 +35,30 @@ type Config struct {
...
+	}
+	writer.total += int64(n)
+	writer.temp += int64(n)
+	// temp holds number of bytes streamed since last push, total bytes streamed
+	// is only pushed once temp reaches threshold (currently 500 megabytes)
+	if writer.temp > 500000000 {
+		writer.tracker.BytesStreamed(writer.total)
+		writer.temp = 0
+	}
+	return n, err
```
Comment (chunmingli) [inline atlas_restore.go]: Can return `n, nil`

File: atlas_restore.go [hunk 7]
```diff
@@ -116,16 +139,32 @@ func writeTarGz(dbDir, tarParentDir string, gzipCompressionLevel int, excludePat
...
+	relPaths := make([]string, 0)
+	size := int64(0)
+
 	iter, err := listFiles(dbDir, ".")
 	if err != nil {
 		return err
 	}
-	for relPath := range iter {
-		if err := addToTar(tw, dbDir, tarParentDir, relPath, excludePaths); err != nil {
+	for path := range iter {
```
Comment (chunmingli) [inline atlas_restore.go]: Total size can probably have its own method and get calculated just once as compared to in every `writeToTarGz` invocation.

File: restore_test.go [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package main
+
+import (
+	"testing"
+)
+
+func TestProgressWriter(t *testing.T) {
+}
```
Comment (chunmingli) [inline restore_test.go]: np: newline

File: status.go (hunk 3 above)
Comment (chunmingli) [reply status.go]: Casing has to do with visibility in Go. See https://golang.org/doc/effective_go.html#names

File: status.go (hunk 5 above)
Comment (chunmingli) [reply status.go]: I think we want to push for both manual download and automated restore, so probably ok to remove shouldPush.

**Comment by kaywux** [comment]
Screenshot of restore job status in restorejobs database:

screen shot 2018-07-27 at 5 37 36 pm

**Comment by chunmingli** [comment]
Remember to set up gofmt for your IDE to automatically format Go code.

File: all_restore_stats.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		DownloadSpeed:	downloadSpeed,
+		BytesStreamed:  bytesStreamed,
+	}
+	return stats
+}
+
+func (p *AllRequestStats) GetRequests() []*RequestStats {
+	requests := make([]*RequestStats, len(p.requestStats))
+	count := 0
+	for _, val := range p.requestStats {
```
Comment (lawctan) [inline all_restore_stats.go]: let's sort by the key before adding to array so that the order is always consistent

File: all_restore_stats.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	req.Status = "Succeeded"
+	req.TimeEnd = time.Now()
+}
+
+func (p *AllRequestStats) Failed(id int64) {
+	req, ok := p.requestStats[id]; if !ok {
+		log.Printf("Request with id %v cannot be found!")
+		return
+	}
+	req.Status = "Failed"
```
Comment (lawctan) [inline all_restore_stats.go]: i think we should add timeEnd here as well. Will be helpful debugging info

File: all_restore_stats.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+type RequestStats struct {
+	Id				int64		`json:"id"`
+	Status			string		`json:"status"`
+	TimeStart		time.Time	`json:"timeStart"`
+	TimeEnd			time.Time	`json:"timeEnd"`
+	DownloadSpeed	string		`json:"downloadSpeed"`
+	BytesStreamed	int64		`json:"bytesStreamed"`
+}
+
+func NewRequestStats(id int64, status string, timeStart time.Time, timeEnd time.Time, downloadSpeed string, bytesStreamed int64) *RequestStats {
```
Comment (lawctan) [inline all_restore_stats.go]: change to uint64 for the 3 params

File: all_restore_stats.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	req, ok := p.requestStats[id]; if !ok {
+		log.Printf("Request with id %v cannot be found!")
+		return
+	}
+	req.Status = "Failed"
+}
+
+func (p *AllRequestStats) AddNewRequest(id int64, status string, timeStart time.Time, timeEnd time.Time, downloadSpeed string, bytesStreamed int64) {
+	if p.requestStats == nil {
+		p.requestStats = make(map[int64]*RequestStats)
```
Comment (lawctan) [inline all_restore_stats.go]: let's put any operations to the map, whether read or write, to their own separate function, and protect it with read write lock (https://golang.org/pkg/sync/#RWMutex).

File: all_restore_stats.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	req := NewRequestStats(id, status, timeStart, timeEnd, downloadSpeed, bytesStreamed)
+	p.requestStats[id] = req
+
+	// if size of map exceeds 100, start clearing failed requests
+	if len(p.requestStats) > 100 {
```
Comment (lawctan) [inline all_restore_stats.go]: let's delete old entries before adding new ones so that we never send more than the capacity.

File: all_restore_stats.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+		p.clearFailedRequests()
+	}
+}
+
+func (p *AllRequestStats) clearFailedRequests() {
+	for key, val := range p.requestStats {
```
Comment (lawctan) [inline all_restore_stats.go]: we only want to delete the oldest failed entry. Not all failed entries. So let's sort the keys first and return right away after the 1st delete

File: all_restore_stats.go [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func (p *AllRequestStats) UpdateRequestProgress(id int64, n int64) {
+	req, ok := p.requestStats[id]
+	if !ok {
+		log.Printf("Request with id %v cannot be found!")
+		return
+	}
+	req.BytesStreamed += n
+	req.DownloadSpeed = fmt.Sprintf("%v bytes per second", float64(n) / time.Now().Sub(req.TimeStart).Seconds())
```
Comment (lawctan) [inline all_restore_stats.go]: i think the dividend should be `float64(req.BytesStreamed)`. And let's store just the value instead of a string

File: atlas_restore.go [hunk 16]
```diff
@@ -30,12 +32,32 @@ type Config struct {
...
+	PushTimeInSeconds	 int	  `yaml:"pushTimeInSeconds"`
 	CentralUrl           string   `yaml:"centralUrl"`
 	JobId                string   `yaml:"jobId"`
 	ExcludePaths         []string `yaml:"excludePaths"`
 }
 
-func addToTar(tw *tar.Writer, dbPath, tarParentDir, relPath string, excludePaths []string) error {
+type ProgressWriter struct {
+	*tar.Writer
+	requestId	int64
```
Comment (lawctan) [inline atlas_restore.go]: uint64

File: atlas_restore.go [hunk 17]
```diff
@@ -30,12 +32,32 @@ type Config struct {
...
+	tracker		*StatsTracker
+}
+
+func (writer *ProgressWriter) Write(p []byte) (int, error) {
+	n, err := writer.Writer.Write(p)
+	if err != nil {
+		return 0, err
+	}
+	writer.tracker.UpdateBytesStreamed(writer.requestId, int64(n))
```
Comment (lawctan) [inline atlas_restore.go]: uint64

File: atlas_restore.go [hunk 18]
```diff
@@ -30,12 +32,32 @@ type Config struct {
...
+	return n, nil
+}
+
+func (writer *ProgressWriter) WriteHeader(header *tar.Header) error {
+	return writer.Writer.WriteHeader(header)
+}
+
+func addToTar(pw *ProgressWriter, dbPath, tarParentDir, relPath string, excludePaths []string) error {
```
Comment (lawctan) [inline atlas_restore.go]: let's change 1st argument to `tw *tar.Writer` since that's more generic. We want it to be more "abstracted", since this function should work for any tar writer. Passing a progressWriter simply modifies the behavior, but using a regular tar.Writer should work normally

File: atlas_restore.go [hunk 19]
```diff
@@ -107,7 +129,28 @@ func listFiles(parent, relDirPath string) (<-chan string, error) {
...
 
-func writeTarGz(dbDir, tarParentDir string, gzipCompressionLevel int, excludePaths []string, w io.Writer) error {
+func returnFilesAndSize(dbDir string) (int64, []string, error) {
+	iter, err := listFiles(dbDir, ".")
+	if err != nil {
+		return int64(0), nil, err
+	}
+
+	relPaths := make([]string, 0)
+	size := int64(0)
```
Comment (lawctan) [inline atlas_restore.go]: uint64

File: atlas_restore.go [hunk 20]
```diff
@@ -107,7 +129,28 @@ func listFiles(parent, relDirPath string) (<-chan string, error) {
...
+			return int64(0), nil, err
+		}
+		size += fileInfo.Size()
+		relPaths = append(relPaths, path)
+	}
+	return size, relPaths, err
+}
+
+func writeTarGz(dbDir, tarParentDir string, gzipCompressionLevel int, 
+	excludePaths []string, w io.Writer, id int64, stats *StatsTracker, relPaths []string) error {
```
Comment (lawctan) [inline atlas_restore.go]: rename `id` to `requestId` for clarity

File: atlas_restore.go [hunk 21]
```diff
@@ -152,16 +193,35 @@ func main() {
...
+	size, relPaths, err := returnFilesAndSize(config.DbPath)
+	if err != nil {
+		return
+	}
+	stats.SetTotalBytes(size)
+
+	idCount := int64(0)
+
+	// create go routine to periodically push progress updates
+	ticker := time.NewTicker(time.Second * time.Duration(config.PushTimeInSeconds))
```
Comment (lawctan) [inline atlas_restore.go]: let's put this ticker logic to a separate function to improve readability

File: atlas_restore.go [hunk 22]
```diff
@@ -152,16 +193,35 @@ func main() {
...
+	go func() {
+		for t := range ticker.C {
+			log.Printf("Pushing progress status at time %s\n", t)
```
Comment (lawctan) [inline atlas_restore.go]: I think we can remove or comment this out since this will be too noisy and doesn't provide much value

File: atlas_restore.go [hunk 23]
```diff
@@ -152,16 +193,35 @@ func main() {
 		config.AuthKey,
 		config.DownloadMode)
 
+	size, relPaths, err := returnFilesAndSize(config.DbPath)
+	if err != nil {
+		return
+	}
+	stats.SetTotalBytes(size)
+
+	idCount := int64(0)
```
Comment (lawctan) [inline atlas_restore.go]: uint64

File: status.go [hunk 24]
```diff
@@ -7,53 +7,56 @@ import (
...
-	jobId      string
-	authKey    string
-	shouldPush bool
+	stats      		DownloadStats
+	requestStats	AllRequestStats
+	centralUrl 		string
+	client     		*http.Client
+	jobId      		string
+	authKey    		string
+	downloadMode   	bool
```
Comment (lawctan) [inline status.go]: seems like we don't use downloadMode anymore in statsTracker, if so, we can remove it

File: atlas_restore.go [hunk 25]
```diff
@@ -30,12 +32,32 @@ type Config struct {
...
 	CentralUrl           string   `yaml:"centralUrl"`
 	JobId                string   `yaml:"jobId"`
 	ExcludePaths         []string `yaml:"excludePaths"`
 }
 
-func addToTar(tw *tar.Writer, dbPath, tarParentDir, relPath string, excludePaths []string) error {
+type ProgressWriter struct {
+	*tar.Writer
+	requestId	int64
+	tracker		*StatsTracker
```
Comment (lawctan) [inline atlas_restore.go]: we can actually store the `requestStats RequestStats` object here as well. And do something like `writer.tracker.UpdateBytesStreamed(requestStats, uint64(n))` to avoid doing a map lookup.

File: atlas_restore.go (hunk 18 above)
Comment (lawctan) [reply atlas_restore.go]: as discussed, we can define an interface with a write() and writeHeader() method signature. But now that I think about it, it's not a big deal. We can stick with `pw *ProgressWriter` too

File: all_restore_stats.go (hunk 12 above)
Comment (kaywux) [reply all_restore_stats.go]: Use regular Mutex Lock/Unlock as discussed

**Comment by kaywux** [comment]
Ready for another look!

File: all_restore_stats.go [hunk 26]
```diff
@@ -2,98 +2,133 @@ package main
...
-		if val.Status == "Failed" {
-			delete(p.requestStats, key)
+func (p *AllRequestStats) clearFailedRequest() {
+	min := ^uint64(0)
+	for _, key := range p.getAllIdsFromMap() {
+		if key < min {
+			min = key
 		}
 	}
+	if min != ^uint64(0) {
```
Comment (lawctan) [inline all_restore_stats.go]: there's the edge case of when the requestIds reach up to the max value. With this `if condition`, it won't delete that entry. So I don't think we need if condition

File: all_restore_stats.go [hunk 27]
```diff
@@ -2,98 +2,133 @@ package main
...
 
-func (p *AllRequestStats) clearFailedRequests() {
-	for key, val := range p.requestStats {
-		if val.Status == "Failed" {
-			delete(p.requestStats, key)
+func (p *AllRequestStats) clearFailedRequest() {
+	min := ^uint64(0)
+	for _, key := range p.getAllIdsFromMap() {
+		if key < min {
+			min = key
```
Comment (lawctan) [inline all_restore_stats.go]: we only want to delete the oldest failed entry right? so you need to add the check for the status as well. And you can do the `sort.Slice` here like you did earlier, and iterate in order.

let's also rename function to `removeOldestFailedRequest`

File: all_restore_stats.go [hunk 28]
```diff
@@ -2,98 +2,133 @@ package main
...
 	"log"
+	"sort"
+	"sync"
 	"time"
-	"fmt"
 )
 
 type AllRequestStats struct {
-	requestStats	map[int64]*RequestStats
+	sync.Mutex
```
Comment (lawctan) [inline all_restore_stats.go]: let's add a comment here on why we chose mutex instead of RWLock

File: all_restore_stats.go [hunk 29]
```diff
@@ -2,98 +2,133 @@ package main
...
 	}
-
-	req := NewRequestStats(id, status, timeStart, timeEnd, downloadSpeed, bytesStreamed)
-	p.requestStats[id] = req
+	p.Unlock()
 
 	// if size of map exceeds 100, start clearing failed requests
-	if len(p.requestStats) > 100 {
-		p.clearFailedRequests()
+	if len(p.getAllIdsFromMap()) > 100 {
```
Comment (lawctan) [inline all_restore_stats.go]: make `100` a const

File: all_restore_stats.go [hunk 30]
```diff
@@ -2,98 +2,133 @@ package main
...
+	p.requestStats[id] = req
+	p.Unlock()
+	return req
 }
 
-func (p *AllRequestStats) clearFailedRequests() {
-	for key, val := range p.requestStats {
-		if val.Status == "Failed" {
-			delete(p.requestStats, key)
+func (p *AllRequestStats) clearFailedRequest() {
```
Comment (lawctan) [inline all_restore_stats.go]: we should move the lock to the beginning of this function, because we don't want 2 threads attempting to delete the same entry

File: all_restore_stats.go [hunk 31]
```diff
@@ -2,98 +2,133 @@ package main
...
 
-func (p *AllRequestStats) InProgress(id int64) {
-	req, ok := p.requestStats[id]; if !ok {
+func (p *AllRequestStats) InProgress(id uint64) {
+	req, ok := p.getRequest(id)
+	if !ok {
 		log.Printf("Request with id %v cannot be found!")
 		return
 	}
 	req.Status = "inProgress"
```
Comment (lawctan) [inline all_restore_stats.go]: let's make these status strings a const at the top level of this file.

File: atlas_restore.go [hunk 32]
```diff
@@ -130,6 +170,16 @@ func writeTarGz(dbDir, tarParentDir string, gzipCompressionLevel int, excludePat
 	return nil
 }
 
+func startProgressPushes(stats *StatsTracker) {
+	const PushTimeInSeconds = 3
```
Comment (lawctan) [inline atlas_restore.go]: perhaps every 5 seconds is better. less frequent and still frequent enough

File: all_restore_stats.go [hunk 33]
```diff
@@ -0,0 +1,139 @@
...
+const MaxNumRequests = 100
+const InProgressString = "InProgress"
+const FailedString = "Failed"
+const SucceededString = "Succeeded"
+
+type AllRequestStats struct {
+	// using Mutex instead of RWlock because read calls to map are fairly
+	//infrequent (on new http request, success or failure)
+	sync.Mutex
+	requestStats map[uint64]*RequestStats
```
Comment (lawctan) [inline all_restore_stats.go]: i think this is confusingly named. RequestStats is a datatype for a single stat, but this is a map so it would contain plenty of values. Something like `requestStatsMap` would be less confusing.

File: status.go [hunk 34]
```diff
@@ -7,53 +7,54 @@ import (
...
-		stats:      DownloadStats{},
-		centralUrl: centralUrl,
-		jobId:      jobId,
-		authKey:    authKey,
-	}
-	if shouldPush {
-		tracker.client = &http.Client{}
-		tracker.shouldPush = true
+		stats:        DownloadStats{},
+		requestStats: AllRequestStats{},
```
Comment (lawctan) [inline status.go]: this will also be less confusing if named `allRequestStats`

File: status.go [hunk 35]
```diff
@@ -75,28 +76,41 @@ func (p *StatsTracker) Push(toPush *DownloadStats) {
...
 	log.Printf("Stats: %v", p.stats)
-	if p.shouldPush {
-		go p.Push(&DownloadStats{InProgress: 1})
-	}
+	p.allRequestStats.InProgress(id)
+	PushProgress(p, 1, 0, 0)
 }
 
-func (p *StatsTracker) Succeeded() {
+func (p *StatsTracker) Succeeded(id uint64) {
```
Comment (chunmingli) [inline status.go]: There are now a pair of `Succeeded` methods, one for `StatsTracker` and one for `AllRequestStats`. I find it hard to follow and confusing.

I'd suggest just having one Succeeded() method on StatsTracker and pass  That way we don't need to do the request id look up from the map. Also update the Succeeded/InProgress counter from there as well. So something like:
```
func (tracker *StatsTracker) Succeeded(req *RequestStats) {
        req.Status = SucceededString
	req.TimeEnd = time.Now()
        tracker.Lock()
        // update the overall Succeeded and Inprogress counter, protected by the mutex
        ...
        tracker.Unlock()
}
```

Similar comments for Failed() and InProgress().

**Comment by chunmingli** [comment]
Oops, looks like it's already merged - but not squash merged. No worries. We can have a separate Jira/PR to address my comments.
